fs/fs.go

```diff
@@ -29,7 +29,7 @@ func NewDatastore(path string) (ds.Datastore, error) {
 
 // KeyFilename returns the filename associated with `key`
 func (d *Datastore) KeyFilename(key ds.Key) string {
-	return filepath.Join(d.path, key.String(), ObjectKeySuffix)
```

Comment:
The real fix is encoding every segment, e.g. hex. Then the extension being non-hex makes namespace collisions impossible. Then dirs don't need an extension.

Keys can also be raw binary, which makes the above necessary anyway.

And the encoding can't be base58 if you care about HFS.

But if it doesn't look pretty in `find`, what's the purpose of preserving the key segments, and not just flattening it (treating the key as opaque binary)? To support faster `Query` prefixes, a feature that isn't used anyway?

For the record, I'm tempted to make the blockstore use a flat version and not this. It's just way simpler and robust, don't have to worry about all the edge cases.